lib/core/test/unit/Cardano/Wallet/Primitive/ScriptsSpec.hs

```diff
@@ -0,0 +1,383 @@
+        it "script composed with not our verification keys should not be discovered"
+            (property prop_scriptFromNotOurVerKeys)
+        it "the same script discovered twice should have the same knownScripts imprint"
+            (property prop_scriptDiscoveredTwice)
+        it "knownScripts of the sequential state is populated properly"
+            (property prop_scriptUpdatesStateProperly)
+        it "scripts with our verification keys are discovered properly"
+            (property prop_scriptsDiscovered)
+        it "discovering our verification keys make them mark Used"
+            (property prop_markingDiscoveredVerKeys)
+        it "discovering works after pool extension"
+            (property prop_poolExtension)
+        it "our verification keys beyond pool gap are not discovered"
+            (property prop_scriptsOutsideNotDiscovered)
+        it "before and after discovery pool number of last consequitive Unused keys stays the same"
+            (property prop_unusedVerKeysConstant)
+        it "discovered verification keys in scripts are consistent between knownScripts and indexedKeys"
+            (property prop_verKeysConsistent)
+
+prop_scriptFromOurVerKeys
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptFromOurVerKeys (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let scriptKeyHashes = retrieveAllVerKeyHashes script
+    let seqState = initializeState accXPub'
+    let (ourSharedKeys, _) = isShared script seqState
+    L.sort (L.nub $ map hashVerificationKey ourSharedKeys) ===
+        L.sort (L.nub scriptKeyHashes)
+
+prop_scriptFromNotOurVerKeys
+    :: ShelleyKey 'AccountK XPub
+    -> AccountXPubWithScripts
+    -> Property
+prop_scriptFromNotOurVerKeys accXPub' (AccountXPubWithScripts _accXPub scripts') = do
+    let (script:_) = scripts'
+    let seqState = initializeState accXPub'
+    let (ourSharedKeys, _) = isShared script seqState
+    ourSharedKeys === []
+
+prop_scriptUpdatesStateProperly
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptUpdatesStateProperly (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let sciptKeyHashes = retrieveAllVerKeyHashes script
+    let seqState = initializeState accXPub'
+    let (_, seqState') = isShared script seqState
+    let scriptKeyHashesInMap =
+            Set.fromList . map (deriveKeyHash accXPub') <$>
+            Map.lookup (toScriptHash script) (getKnownScripts seqState')
+    scriptKeyHashesInMap === Just (Set.fromList (L.nub sciptKeyHashes))
+
+prop_scriptDiscoveredTwice
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptDiscoveredTwice (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let seqState = initializeState accXPub'
+    let (_, seqState') = isShared script seqState
+    let (_, seqState'') = isShared script seqState'
+    getKnownScripts seqState' === getKnownScripts seqState''
```

Comment:
I'd even add that there shouldn't be any new key discovered as well. Basically, `seqState' === seqState''` :)

Reply:
done done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/ab8567640737c089cb9f15c4be3cfce6dbcadb1a